### PR TITLE
BUG: Fix decref before incref for in-place accumulate

### LINKED ISSUE #7402: cumsum of mutable objects in an array gives wrong results
While trying to replace Python `long` integers with gmpy2 `mpz` objects in an `object` array, I discovered a rather strange behavior of the *in-place* `np.cumsum` function (with `out=`). It seems that the objects are reused in some way. E.g.
```python
import numpy as np
from gmpy2 import mpz

arr = np.empty(4, dtype=object)
arr[:] = [mpz(1) for _ in range(4)]
np.cumsum(arr, out=arr)
# np.cumsum(arr, out=arr)
print arr
```
prints `[mpz(2) mpz(2) mpz(3) mpz(4)]`.

With the second `cumsum` uncommented, the values are `7, 7, 7, 11`. There is no problem if `long` is used instead of `mpz`.

It seems that a similar issue can be reproduced using tuples and lists in an array.

```python
import numpy as np

arr = np.empty(4, dtype=object)
arr[:] = [(1,) for _ in range(4)]
np.cumsum(arr, out=arr)
print arr  
arr[:] = [[1] for _ in range(4)]
np.cumsum(arr, out=arr)
print arr
```

As I would expect, the first print shows an array of tuples of growing length. However, the second print causes `Segmentation fault: 11`.

### PR DESCRIPTION
**Pull request by seberg**
closes gh-7402

### CODE REVIEW

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -3239,17 +3239,16 @@ PyUFunc_Accumulate(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *out,
...
         do {
-
             dataptr_copy[0] = dataptr[0];
             dataptr_copy[1] = dataptr[1];
             dataptr_copy[2] = dataptr[0];
 
             /* Copy the first element to start the reduction */
             if (otype == NPY_OBJECT) {
+                Py_XINCREF(*(PyObject **)dataptr_copy[1]);
                 Py_XDECREF(*(PyObject **)dataptr_copy[0]);
```
Comment (njsmith) [inline numpy/core/src/umath/ufunc_object.c]: So to check if I understand: what's going on here is, we're doing an accumulate, `dataptr_copy[0]` points to the `out` array, `dataptr_copy[1]` points to the `in` array, and when `out is in` then bad stuff can happen. In that case I think this logic is now correct, though it is very tricky :-).

Observation 1: this desperately needs a comment :-).

Observation 2: could we instead replace this whole `if/else` stuff with a call to the `dtype`'s data copy method? that would let us remove the special case, we'll eventually need it when dtypes get smarter, and object dtype method should already handle this case?

**Comment by charris** [comment]
LGTM, but agree with @njsmith that a comment would be useful.

**Comment by charris** [comment]
@seberg I want to put this in an backport for 1.11 (obviously correct), but a comment as to why the order is important would be very helpful.

**Comment by seberg** [comment]
Added a (maybe a bit large) comment. I guess there is a point, Nathaniel is right about replacing these things probably, though maybe it would be better to replace more then just in this one place then as  well.
Anyway, for the sake of backporting, no further changes.

**Comment by charris** [comment]
Looks good pending AppVeyor. A shorter comment would be something like
```
/*
 * dataptr[0] and dataptr[1] may point to the same object,
 * so Py_INCREF before Py_DECREF.
 */
```

**Comment by njsmith** [comment]
I would expand "need to incref" to "need to incref before decref", but I like the current comment, it seems to give the right amount of information to me and we're not really in danger of leaving this code *over*-commented :-)

**Comment by seberg** [comment]
Heh, yeah, at the very least a "first" is missing, will be added in a minute or two ;)

**Comment by seberg** [comment]
OK, just left the long comment for now, true, Chucks version is nice too ;).

**Comment by charris** [comment]
Thanks Sebastian. Nice find also.

**Pull request merged**